File: src/main/java/org/stt/gui/jfx/JFXModule.java
package org.stt.gui.jfx;

import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoSet;
import dagger.multibindings.Multibinds;
import javafx.scene.control.Hyperlink;
import javafx.scene.text.Font;
import javafx.scene.text.Text;
import javafx.stage.Stage;
import org.stt.config.ActivitiesConfig;
import org.stt.text.ItemGrouper;

import javax.inject.Named;
import javax.inject.Singleton;
import java.awt.*;
import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.net.URI;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

@Module
public abstract class JFXModule {
    private static final Pattern URL_PATTERN = Pattern.compile("\\b(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]");

    private JFXModule() {
    }

    @Singleton
    @Provides
    static Stage stageProvider() {
        return new Stage();
    }

    @Multibinds
    abstract Set<ActivityTextDisplayProcessor> activityMappers();

    @Provides
    @Named("activityToText")
    static ActivityTextDisplayProcessor provideActivityToTextMapper(Set<ActivityTextDisplayProcessor> mappers) {
        return o -> {
            Stream<Object> stream = o;
            for (ActivityTextDisplayProcessor mapper : mappers) {
                stream = mapper.apply(stream);
            }
            return stream;
        };
    }

    @Provides
    @IntoSet
    static ActivityTextDisplayProcessor groupingMapper(ActivitiesConfig activitiesConfig, ItemGrouper grouper) {

        return in -> {
            Boolean[] first = {true};
            return in.flatMap(o -> {
                if (!first[0]) {
                    return Stream.of(o);
                }
                first[0] = false;
                if (o instanceof String && activitiesConfig.isGroupItems()) {
                    return dissect(grouper, (String) o);
                }
                return Stream.of(o);
            });
        };
    }

    private static Stream<?> dissect(ItemGrouper grouper, String o) {
        List<Object> result = new ArrayList<>();
        List<ItemGrouper.Group> groups = grouper.getGroupsOf(o);
        for (int i = 0; i < groups.size(); i++) {
            if (i > 0) {
                result.add(" ");
            }
            ItemGrouper.Group group = groups.get(i);
            boolean last = i >= groups.size() - 1;
            if (last) {
                result.add(group.content);
            } else {
                Text text = new Text(group.content);
                text.getStyleClass().add("reportGroup" + i);
                result.add("\u2768");
                result.add(text);
                result.add("\u2769");
            }
        }
        return result.stream();
    }

    @Provides
    @IntoSet
    static ActivityTextDisplayProcessor hyperlinkMapper(ExecutorService executorService) {
        return in -> in.flatMap(o -> {
            if (o instanceof String) {
                return toHyperlink(executorService, (String) o);
            }
            return Stream.of(o);
        });
    }

    private static Stream<?> toHyperlink(ExecutorService executorService, String activity) {
        List<Object> result = new ArrayList<>();
        Matcher matcher = URL_PATTERN.matcher(activity);
        int index = 0;
        while (matcher.find(index)) {
            String preamble = activity.substring(index, matcher.start());
            if (!preamble.isEmpty()) {
                result.add(preamble);
            }
            String uri = activity.substring(matcher.start(), matcher.end());
            Hyperlink hyperlink = new Hyperlink(uri);
            hyperlink.setOnAction(event -> executorService.submit(() -> {
                try {
                    Desktop.getDesktop().browse(URI.create(uri));
                } catch (IOException e) {
                    throw new UncheckedIOException(e);
                }
            }));
            result.add(hyperlink);
            index = matcher.end();
        }
        result.add(activity.substring(index));
        return result.stream();
    }

    @Provides
    @Named("glyph")
    static Font provideFont() {
        try (InputStream fontStream = JFXModule.class.getResourceAsStream("/fontawesome-webfont.ttf")) {
            return Font.loadFont(fontStream, 0);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: src/main/kotlin/org/stt/gui/jfx/JFXModule.kt
package org.stt.gui.jfx

import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoSet
import dagger.multibindings.Multibinds
import javafx.scene.control.Hyperlink
import javafx.scene.text.Font
import javafx.scene.text.Text
import javafx.stage.Stage
import org.stt.config.ActivitiesConfig
import org.stt.text.ItemGrouper
import java.awt.Desktop
import java.io.IOException
import java.io.UncheckedIOException
import java.net.URI
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.regex.Pattern
import java.util.stream.Stream
import javax.inject.Named
import javax.inject.Singleton

private val URL_PATTERN = Pattern.compile("\\b(https?|ftp|file)://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]")

typealias ActivityTextDisplayProcessor = (Stream<Any>) -> Stream<Any>

@Module
abstract class JFXModule {

    @Multibinds
    abstract fun activityMappers(): Set<ActivityTextDisplayProcessor>

    @Module
    companion object {

        @Singleton
        @Provides
        @JvmStatic
        fun stageProvider(): Stage {
            return Stage()
        }

        @Provides
        @Named("activityToText")
        @JvmStatic
        fun provideActivityToTextMapper(mappers: Set<@JvmSuppressWildcards ActivityTextDisplayProcessor>): ActivityTextDisplayProcessor =
                {
                    var stream = it
                    for (mapper in mappers) {
                        stream = mapper(stream)
                    }
                    stream
                }

        @Provides
        @IntoSet
        @JvmStatic
        fun groupingMapper(activitiesConfig: ActivitiesConfig, grouper: @JvmSuppressWildcards ItemGrouper): ActivityTextDisplayProcessor =
                {
                    var first = true
                    it.flatMap {
                        if (!first) {
                            return@flatMap Stream.of(it)
                        }
                        first = false
                        if (it is String && activitiesConfig.isGroupItems) {
                            return@flatMap dissect(grouper, it)
                        }
                        Stream.of(it)
                    }
                }

        private fun dissect(grouper: ItemGrouper, o: String): Stream<*> {
            val result = ArrayList<Any>()
            val groups = grouper(o)
            for (i in groups.indices) {
                if (i > 0) {
                    result.add(" ")
                }
                val group = groups[i]
                val last = i >= groups.size - 1
                if (last) {
                    result.add(group.content)
                } else {
                    val text = Text(group.content)
                    text.styleClass.add("reportGroup$i")
                    result.add("\u2768")
                    result.add(text)
                    result.add("\u2769")
                }
            }
            return result.stream()
        }

        @Provides
        @IntoSet
        @JvmStatic
        fun hyperlinkMapper(executorService: ExecutorService): ActivityTextDisplayProcessor =
                {
                    it.flatMap {
                        return@flatMap if (it is String) {
                            toHyperlink(executorService, it)
                        } else {
                            Stream.of<Any>(it)
                        }
                    }
                }

        private fun toHyperlink(executorService: ExecutorService, activity: String): Stream<*> {
            val result = ArrayList<Any>()
            val matcher = URL_PATTERN.matcher(activity)
            var index = 0
            while (matcher.find(index)) {
                val preamble = activity.substring(index, matcher.start())
                if (!preamble.isEmpty()) {
                    result.add(preamble)
                }
                val uri = activity.substring(matcher.start(), matcher.end())
                val hyperlink = Hyperlink(uri)
                hyperlink.setOnAction {
                    executorService.submit {
                        try {
                            Desktop.getDesktop().browse(URI.create(uri))
                        } catch (e: IOException) {
                            throw UncheckedIOException(e)
                        }
                    }
                }
                result.add(hyperlink)
                index = matcher.end()
            }
            result.add(activity.substring(index))
            return result.stream()
        }

        @Provides
        @Named("glyph")
        @JvmStatic
        fun provideFont(): Font {
            try {
                JFXModule::class.java.getResourceAsStream("/fontawesome-webfont.ttf").use { fontStream -> return Font.loadFont(fontStream, 0.0) }
            } catch (e: IOException) {
                throw UncheckedIOException(e)
            }

        }
    }
}
